// gpthandler/source/types.ts

type MediaContent = { image: string };
type TextContent = string;
export type Role = 'user' | 'assistant' | 'system';
export type MessageContent = TextContent | { text: TextContent, media: MediaContent[] };
export type Message = { role: Role, content: MessageContent };
export type ChatHistory = Message[];

export type ModelData = { method: string, model: string };
type OpenAIModel = { method: 'OpenAi', model: 'gpt-3.5-turbo-1106' } | { method: 'OpenAi', model: 'gpt-4-1106-preview' };
type Gpt4FreeModel = { method: 'G4f', model: 'gpt-4-32k' } | { method: 'G4f', model: 'gpt-3.5-turbo-16k' };
type LlmModel = OpenAIModel | Gpt4FreeModel;

type LlmModelsNameEnum = {
    "OpenAi": { gpt4: { method: 'OpenAi', model: 'gpt-4-1106-preview' }, gpt3_5_t: { method: 'OpenAi', model: 'gpt-3.5-turbo-1106' } },
    "G4f": { gpt4: { method: 'G4f', model: 'gpt-4-32k' }, gpt3_5_t: { method: 'G4f', model: 'gpt-3.5-turbo-16k' } },
}
export const llmModelsEnum: LlmModelsNameEnum = {
    "OpenAi": { gpt4: { method: 'OpenAi', model: 'gpt-4-1106-preview' }, gpt3_5_t: { method: 'OpenAi', model: 'gpt-3.5-turbo-1106' } },
    "G4f": { gpt4: { method: 'G4f', model: 'gpt-4-32k' }, gpt3_5_t: { method: 'G4f', model: 'gpt-3.5-turbo-16k' } },
}

const test = llmModelsEnum.G4f.gpt3_5_t;

export interface LlmMethod {
    requestChat(history: ChatHistory, model: string): Promise<LlmAns>;
    translate(text: string, from: string, to: string): Promise<string>;
}

export type ErrorHandlingMode = "as responce text" | "throw error";

export type LlmOptions = { retryCount: number, llmModel: ModelData, errorHandlingMode: ErrorHandlingMode };
export type LlmAns = string;

export type Checker = (arg: any) => boolean

export type TranslationResult = {
    source: { code: string, lang: string },
    target: { code: string, lang: string },
    translation: { parts: any[], result: string }
};

export const langCodes = [
    { id: 'es', name: 'Español' },
    { id: 'en', name: 'English' },
    { id: 'af', name: 'Afrikaans' },
    { id: 'ay', name: 'Aymara' },
    { id: 'sq', name: 'Shqip' },
    { id: 'de', name: 'Deutsch' },
    { id: 'am', name: 'አማርኛ' },
    { id: 'ar', name: 'العربية' },
    { id: 'hy', name: 'Հայերեն' },
    { id: 'as', name: 'অসমীয়া' },
    { id: 'az', name: 'Azərbaycan' },
    { id: 'bm', name: 'Bamanankan' },
    { id: 'bn', name: 'বাংলা' },
    { id: 'bho', name: 'भोजपुरी' },
    { id: 'be', name: 'Беларуская' },
    { id: 'my', name: 'မြန်မာ' },
    { id: 'bs', name: 'Bosanski' },
    { id: 'bg', name: 'Български' },
    { id: 'km', name: 'ភាសាខ្មែរ' },
    { id: 'kn', name: 'ಕನ್ನಡ' },
    { id: 'ca', name: 'Català' },
    { id: 'ceb', name: 'Cebuano' },
    { id: 'cs', name: 'Čeština' },
    { id: 'ny', name: 'Chichewa' },
    { id: 'zh-CN', name: '简体中文' },
    { id: 'zh-TW', name: '繁體中文' },
    { id: 'si', name: 'සිංහල' },
    { id: 'ko', name: '한국어' },
    { id: 'co', name: 'Corsu' },
    { id: 'ht', name: 'Kreyòl Ayisyen' },
    { id: 'hr', name: 'Hrvatski' },
    { id: 'da', name: 'Dansk' },
    { id: 'dv', name: 'ދިވެހި' },
    { id: 'doi', name: 'डोगरी' },
    { id: 'sk', name: 'Slovenčina' },
    { id: 'sl', name: 'Slovenščina' },
    { id: 'eo', name: 'Esperanto' },
    { id: 'et', name: 'Eesti' },
    { id: 'eu', name: 'Euskara' },
    { id: 'ee', name: 'Eʋegbe' },
    { id: 'fi', name: 'Suomi' },
    { id: 'fr', name: 'Français' },
    { id: 'fy', name: 'Frysk' },
    { id: 'gd', name: 'Gàidhlig' },
    { id: 'cy', name: 'Cymraeg' },
    { id: 'gl', name: 'Galego' },
    { id: 'ka', name: 'ქართული' },
    { id: 'el', name: 'Ελληνικά' },
    { id: 'gn', name: 'Avañe\'ẽ' },
    { id: 'gu', name: 'ગુજરાતી' },
    { id: 'ha', name: 'Hausa' },
    { id: 'haw', name: 'Hawaiʻi' },
    { id: 'iw', name: 'עברית' },
    { id: 'hi', name: 'हिन्दी' },
    { id: 'hmn', name: 'Hmoob' },
    { id: 'hu', name: 'Magyar' },
    { id: 'ig', name: 'Igbo' },
    { id: 'ilo', name: 'Ilokano' },
    { id: 'id', name: 'Bahasa Indonesia' },
    { id: 'ga', name: 'Gaeilge' },
    { id: 'is', name: 'Íslenska' },
    { id: 'it', name: 'Italiano' },
    { id: 'ja', name: '日本語' },
    { id: 'jw', name: 'Basa Jawa' },
    { id: 'kk', name: 'Қазақша' },
    { id: 'rw', name: 'Kinyarwanda' },
    { id: 'ky', name: 'Кыргызча' },
    { id: 'gom', name: 'कोंकणी' },
    { id: 'kri', name: 'Krio' },
    { id: 'ku', name: 'Kurdî' },
    { id: 'ckb', name: 'کوردیی ناوەندی' },
    { id: 'lo', name: 'ລາວ' },
    { id: 'la', name: 'Latina' },
    { id: 'lv', name: 'Latviešu' },
    { id: 'ln', name: 'Lingála' },
    { id: 'lt', name: 'Lietuvių' },
    { id: 'lg', name: 'Luganda' },
    { id: 'lb', name: 'Lëtzebuergesch' },
    { id: 'mk', name: 'Македонски' },
    { id: 'mai', name: 'मैथिली' },
    { id: 'ml', name: 'മലയാളം' },
    { id: 'ms', name: 'Bahasa Melayu' },
    { id: 'mg', name: 'Malagasy' },
    { id: 'mt', name: 'Malti' },
    { id: 'mi', name: 'Māori' },
    { id: 'mr', name: 'मराठी' },
    { id: 'mni-Mtei', name: 'ꯃꯤꯇꯩ ꯂꯣꯟ' },
    { id: 'lus', name: 'Mizo ṭawng' },
    { id: 'mn', name: 'Монгол' },
    { id: 'nl', name: 'Nederlands' },
    { id: 'ne', name: 'नेपाली' },
    { id: 'no', name: 'Norsk' },
    { id: 'or', name: 'ଓଡ଼ିଆ' },
    { id: 'om', name: 'Oromoo' },
    { id: 'pa', name: 'ਪੰਜਾਬੀ' },
    { id: 'ps', name: 'پښتو' },
    { id: 'fa', name: 'فارسی' },
    { id: 'pl', name: 'Polski' },
    { id: 'pt', name: 'Português' },
    { id: 'qu', name: 'Runa Simi' },
    { id: 'ro', name: 'Română' },
    { id: 'ru', name: 'Русский' },
    { id: 'sm', name: 'Gagana Samoa' },
    { id: 'sa', name: 'संस्कृतम्' },
    { id: 'nso', name: 'Sesotho sa Leboa' },
    { id: 'sr', name: 'Српски' },
    { id: 'st', name: 'Sesotho' },
    { id: 'sn', name: 'ChiShona' },
    { id: 'sd', name: 'سنڌي' },
    { id: 'so', name: 'Soomaali' },
    { id: 'sw', name: 'Kiswahili' },
    { id: 'sv', name: 'Svenska' },
    { id: 'su', name: 'Basa Sunda' },
    { id: 'tl', name: 'Filipino' },
    { id: 'th', name: 'ไทย' },
    { id: 'ta', name: 'தமிழ்' },
    { id: 'tt', name: 'Татар' },
    { id: 'tg', name: 'Тоҷикӣ' },
    { id: 'te', name: 'తెలుగు' },
    { id: 'ti', name: 'ትግርኛ' },
    { id: 'ts', name: 'Xitsonga' },
    { id: 'tr', name: 'Türkçe' },
    { id: 'tk', name: 'Türkmençe' },
    { id: 'ak', name: 'Akan' },
    { id: 'uk', name: 'Українська' },
    { id: 'ug', name: 'ئۇيغۇرچە' },
    { id: 'ur', name: 'اردو' },
    { id: 'uz', name: 'Oʻzbekcha' },
    { id: 'vi', name: 'Tiếng Việt' },
    { id: 'xh', name: 'isiXhosa' },
    { id: 'yi', name: 'ייִדיש' },
    { id: 'yo', name: 'Yorùbá' },
    { id: 'zu', name: 'isiZulu' }
];
